### ISSUE TITLE
Class definition for "MyClass" depends on itself

### ISSUE THREAD

**Issue by matt-the-midnight-hacker** [microsoft/pyright#4518]
Note: if you are reporting a wrong signature of a function or a class in the standard library, then the typeshed tracker is better suited for this report: https://github.com/python/typeshed/issues.

**Describe the bug**
At seemingly random times I get errors reported for my classes saying "Class definition for "MyClass" depends on itself Pylance[reportGeneralTypeIssues]"
Also all usages of the class are then flagged with error "Cannot access member "my_member" for type "MyClass"
  Member "my_member" is unknown". When I then save the file containing the class so that pyright gets again invoked the error goes away / or not.. I'm using vscode. I saw it happen on classes using `attrs` library and on slotted classes, not sure if it also happen on other class types or if this has even to do with that.

**To Reproduce**
I don't know, it happens to many different classes at any time.

**Expected behavior**
The error should not happen or if there is really a issue with my code than the error should stay as long as I don't fix anything.

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
![image](https://user-images.githubusercontent.com/83934112/214211224-78c2c58a-7ade-4c0a-82fe-2031680045f6.png)

If your code relies on symbols that are imported from a third-party library, include the associated import statements and specify which versions of those libraries you have installed.

**VS Code extension or command-line**
Are you running pyright as a VS Code extension or a command-line tool? Which version? You can find the version of the VS Code extension by clicking on the Pyright icon in the extensions panel.

I'm using the Pylance v2023.1.30 vscode extension. Extension is enabled on 'WSL: Ubuntu-18.04' if that matters.

**Comment by erictraut** [comment]
Thanks for reporting the problem. Without a repro case, there's not much we can do to fix it. Are you able to replicate the problem with a single, self-contained source file? 

This could be the same underlying problem reported in [this issue](https://github.com/microsoft/pyright/issues/4513). Please see the explanation and recommended steps I posted there.

**Comment by matt-the-midnight-hacker** [comment]
I tried a bit around and unfortunately I'm not able to reproduce the problem with a single, self-contained source file. It happens for me in a bigger project in a complex class hierarchy. I'll let you know once I can see a way to reproduce it.

**Comment by erictraut** [comment]
OK, thanks for trying to repro.

Without a repro case, there's not much I can do about it, so I'm going to close the issue. If you find a way to repro, I'm happy to reopen and investigate further.

**Comment by gakhov** [comment]
I have the same behavior which was gone after I restart Pylance.

![Screenshot 2023-06-20 at 09 23 27](https://github.com/microsoft/pyright/assets/929775/3c61172c-97b3-4c27-a539-11061ccc0015)

And this is the referenced class
![Screenshot 2023-06-20 at 09 23 45](https://github.com/microsoft/pyright/assets/929775/0bcc5acd-29c4-454b-b9f4-2eeb8acbae53)

**Comment by erictraut** [comment]
@gakhov, if you have a repro case, please provide a minimal self-contained code sample in text form. I won't be able to repro it from partial screen shots.

**Comment by gakhov** [comment]
Unfortunately, it is the same for me. I cannot reliably reproduce it: it just happens and has gone with reload, sorry ...

**Comment by LeandroBarone** [comment]
Same issue here. It happens at random: each time I save there's about a 1% chance of some random class breaking. Right now I'm seeing it for datetime.now():

![image](https://github.com/microsoft/pyright/assets/8425029/4b1da3c2-9889-49b1-a224-68d5afa761c8)

**Comment by erictraut** [comment]
Based on the symptom, this sounds like a problem related to evaluation order. Pyright is able to evaluate the type of any identifier in any order. It recursively evaluates dependent types until it's able to determine the type of the requested identifier. It caches any types that it evaluates along the way. It also stashes away any errors that it finds. In theory, it shouldn't matter which order evaluation occurs, but bugs can produce different types based on different orders. Circular dependencies between types can be especially problematic because there's no well-defined order to a cycle.

One technique I use when investigating problems involving evaluation ordering is to try to influence the order by moving the mouse pointer over different identifiers immediately after modifying the file (e.g. by typing a space somewhere innocuous). When you modify a file, pyright immediately discards all of its cached type information for that file and then starts a 250ms timer to re-evaluate the types of all identifiers in the file starting from the top. If you type again within that 250ms window, it resets the timer. This allows you to quickly type multiple characters without needlessly analyzing the file. If, during that 250ms window, you jiggle your mouse pointer over an identifier, it will result in the hover provider (the component that produces the "hover text") asking pyright to evaluate the type for that identifier. Since the type cache is empty at that point, pyright will immediately re-evaluate the type for that identifier at that position in the code. Using this technique, I can often find a way to reliably repro otherwise-sporadic errors. 

Give that a try next time you hit this problem. Any clues you find would be most appreciated.

**Comment by CJTurpie** [comment]
I am also experiencing this regularly (multiple times a day). I also have a repro case as it seems to need a suitably complex repo for it to happen. I tried the trick you suggested with hovering the mouse after adding a space and this normally resolves the issue for that file/class but obviously it would be better if it didn't happen in the first place. Seeing as it happens so often for me is there anything I can do to help to try to diagnose the issue (e.g. diagnostics or logs)?

**Comment by erictraut** [comment]
I need a repro case. If you have a deterministic way to repro it, please provide instructions.

**Comment by cainmagi** [comment]
I got the same error. It randomly happens on random files when I work on a large project, which means that this issue does not happen with specific codes. For some confidentiality reasons, I cannot provide you with my whole project, but I can show you what the issue looks like:
<img width="613" alt="definition-on-itself" src="https://github.com/microsoft/pyright/assets/32186723/e46bf17a-fa1a-46a8-bdb2-63f69186d568">

When this issue happen, it will raise a lot of "fake" errors, which make `pyright` very slow

```logs
2023-07-13 21:32:57.981 [info] [Info  - 9:32:57 PM] (11460) Pylance language server 2023.7.20 (pyright 74d8f3c4) starting
2023-07-13 21:32:57.981 [info] [Info  - 9:32:57 PM] (11460) Server root directory: c:\Users\Yuchen Jin\.vscode\extensions\ms-python.vscode-pylance-2023.7.20\dist
2023-07-13 21:32:57.989 [info] [Info  - 9:32:57 PM] (11460) Starting service instance "...(project-name)"
2023-07-13 21:32:58.065 [info] [Info  - 9:32:58 PM] (11460) Setting pythonPath for service "...(project-name)": "C:\Users\Yuchen Jin\.conda\envs\...\python.exe"
2023-07-13 21:32:58.074 [info] [Warn  - 9:32:58 PM] (11460) stubPath c:\...\typings is not a valid directory.
2023-07-13 21:32:58.257 [info] [Info  - 9:32:58 PM] (11460) Assuming Python version 3.8
2023-07-13 21:32:59.578 [info] [Info  - 9:32:59 PM] (11460) Found 290 source files
2023-07-13 21:32:59.988 [info] [Info  - 9:32:59 PM] (11460) Background analysis(1) root directory: c:\Users\Yuchen Jin\.vscode\extensions\ms-python.vscode-pylance-2023.7.20\dist
2023-07-13 21:32:59.989 [info] [Info  - 9:32:59 PM] (11460) Background analysis(1) started
2023-07-13 21:33:00.193 [info] [Info  - 9:33:00 PM] (11460) Indexer background runner(2) root directory: c:\Users\Yuchen Jin\.vscode\extensions\ms-python.vscode-pylance-2023.7.20\dist (index)
2023-07-13 21:33:00.193 [info] [Info  - 9:33:00 PM] (11460) Indexing(2) started
2023-07-13 21:33:02.611 [info] [Info  - 9:33:02 PM] (11460) [IDX(2)] Long operation: scan packages c:\... (2413ms)
2023-07-13 21:33:02.611 [info] [Info  - 9:33:02 PM] (11460) scanned(2) 301 files over 1 exec env
2023-07-13 21:33:04.270 [info] [Info  - 9:33:04 PM] (11460) indexed(2) 220 files over 1 exec env
2023-07-13 21:33:04.355 [info] [Info  - 9:33:04 PM] (11460) Indexing finished(2).
2023-07-13 21:48:12.932 [info] [Info  - 9:48:12 PM] (11460) [BG(1)] Long operation: checking: c:\...\lazynpio.py (4044ms)
2023-07-13 21:48:12.932 [info] [Info  - 9:48:12 PM] (11460) [BG(1)] Long operation: analyzing: c:\...\lazynpio.py (4179ms)
2023-07-14 08:08:45.762 [info] [Info  - 8:08:45 AM] (11460) Found 291 source files
2023-07-14 08:11:09.865 [info] [Info  - 8:11:09 AM] (11460) [BG(1)] Long operation: getSemanticTokens full at c:\...\backends\extraction\interpolate.py (35800ms)
2023-07-14 08:11:14.140 [info] [Info  - 8:11:14 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (2388ms)
2023-07-14 08:11:14.140 [info] [Info  - 8:11:14 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (2388ms)
2023-07-14 08:11:20.401 [info] [Info  - 8:11:20 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (2737ms)
2023-07-14 08:11:20.401 [info] [Info  - 8:11:20 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (2737ms)
2023-07-14 08:11:25.435 [info] [Info  - 8:11:25 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (4909ms)
2023-07-14 08:11:25.435 [info] [Info  - 8:11:25 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (4909ms)
2023-07-14 08:11:30.523 [info] [Info  - 8:11:30 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (4460ms)
2023-07-14 08:11:30.523 [info] [Info  - 8:11:30 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (4460ms)
2023-07-14 08:11:37.739 [info] [Info  - 8:11:37 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (7213ms)
2023-07-14 08:11:37.739 [info] [Info  - 8:11:37 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (7213ms)
2023-07-14 08:11:48.823 [info] [Info  - 8:11:48 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (2543ms)
2023-07-14 08:11:48.823 [info] [Info  - 8:11:48 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (2543ms)
2023-07-14 08:11:57.235 [info] [Info  - 8:11:57 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (2455ms)
2023-07-14 08:11:57.235 [info] [Info  - 8:11:57 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (2455ms)
2023-07-14 08:12:46.653 [info] [Info  - 8:12:46 AM] (11460) [BG(1)] Long operation: checking: c:\...\backends\extraction\interpolate.py (31153ms)
2023-07-14 08:12:46.653 [info] [Info  - 8:12:46 AM] (11460) [BG(1)] Long operation: analyzing: c:\...\backends\extraction\interpolate.py (31153ms)
2023-07-14 08:32:18.216 [info] [Info  - 8:32:18 AM] (11460) [BG(1)] Long operation: getSemanticTokens delta previousResultId:1689340234064 at c:\...\backends\extraction\interpolate.py (34271ms)
```

For confidentiality reasons, I have replaced the information related to my project name with `...`.

This issue will disappear if I relaunch Code.

**Comment by erictraut** [comment]
@cainmagi, I understand the symptoms. Unfortunately, I can't debug issues from partial screen shots. Could you create a minimal, self-contained code sample (e.g. with function names changed to hide any confidential information) that demonstrates the problem?

**Comment by cainmagi** [comment]
> @cainmagi, I understand the symptoms. Unfortunately, I can't debug issues from partial screen shots. Could you create a minimal, self-contained code sample (e.g. with function names changed to hide any confidential information) that demonstrates the problem?

That may not be easy, because the codes are also confidential. Anyway, I will try to write some non-confidential dummy testing codes to reproduce this issue. Since this issue happens randomly, I am not sure how long it will take for me to do that. This may take a very long period, so please do not wait for me. Once I spot the same issue with these "dummy" codes, I will provide it to you. 

Thank you!

**Comment by erictraut** [comment]
The `classDefinitionCycle` is an interesting clue. That probably means that you have a class definition that relies on itself in a cyclical manner that pyright does not handle. Refer to [this documentation](https://microsoft.github.io/pyright/#/mypy-comparison?id=circular-references) for some examples of this.

**Comment by danstewart** [comment]
Thanks for the info, but the class doesn't depend on itself (at least from what I can tell, and if it does the issue is the opposite, why does the warning not always show).

Pyright/pylance just sometimes thinks it does, if you edit something to retrigger analysis or restart VS code the warning goes away.

I was trying to find the pyright logic that throws this warning but I can't find where `classDefinitionCycle` is actually used.

Since pylance runs after each edit is it possibly a race condition where a later edit finishes before the first edit?
eg. 
1. Make an edit, the code is bad
2. Pylance starts analysis 1
3. Make another edit
4. Pylance starts analysis 2
5. Analysis 2 finishes, everything is fine
6. Analysis 1 finishes, there is an issue

Given all reports (so far) are using pylance should this issue be re-opened under the pylance repo?

Related: https://github.com/microsoft/pylance-release/discussions/4599

**Comment by erictraut** [comment]
Pyright does not analyze your code from top to bottom. The analysis order can be affected by things like hover requests (where you're mousing over an identifier shortly after modifying a file) or completion requests (if you're actively typing). In most cases, this is fine because pyright can determine a a definitive evaluation ordering. If it's asked to evaluate symbol `A`, and that depends on `B` which depends on `C`, it can resolve those in dependency order. However, in cases where there's a cyclical dependency in types, it needs to arbitrarily "break the cycle", and this can result in different results depending on where it entered the cycle. These sort of cycles are notoriously difficult to detect. They often happen to produce correct type evaluations but in rare conditions may not, so reporting a cycle 100% of the time may be seen as a false positive. If this is leading to inconsistent behavior for many users, I may need to rethink how I handle this.

The `classDefinitionCycle` error message is reported in the case where pyright is evaluating a member access expression (e.g. `x.y`) where `x` depends on a class that is "partially evaluated" — that is, pyright is still evaluating the `class` statement associated with that class. I can't think of an example in a working ".py" file because such cycles result in runtime errors. It's possible to create such cycles in a stub file.

Here's an example that someone created in a stub file. The expression `Colors.V` is used in a base class for `_Colors`, but `Colors` uses `_Colors` as a metaclass. There's no way for a type checker to resolve this cycle because the behavior of a member access expression can be impacted by the metaclass.

```python
from typing import Generic, NewType, TypeVar

T = TypeVar("T")

class _EnumTypeWrapper(Generic[T]): ...

class Colors(metaclass=_Colors):
    V = NewType("V", int)

class _Colors(_EnumTypeWrapper[Colors.V], type):
    RED = Colors.V(0)
    GREEN = Colors.V(1)
    BLUE = Colors.V(2)
```

**Comment by giladpn** [comment]
@erictraut I am also suffering from this issue, and on a regular basis. My "procedure" for reproducing this is rather simple, but as you will see, it cannot be boiled down to 3 lines of code. It is however "deterministic" in the sense that it happens on a completely regular basis.

1. do this on a large project with several .ipynb and .py files and using several packages you have written yourself (in .py files), as well as multiple imports from PyPi standard packages and e.g. scikit-learn. The bit of using (from the .ipynb files) the packages you have written yourself (in .py files in separate folders) seems to be quite important. A specific layout of packages is below
2. Also, in my case at least one of the packages is imported into the .ipynb files using a `sys.path.append`
(note `mybeloved/src` in the example below contains a subfolder `mybeloved/src/yodapacks`)
```
sys.path.append(os.path.abspath('../mybeloved/src')+'/')
from yodapacks.pack1 import xcode
from yodapacks.pack2 import ycode
from yodapacks.pack3 import zcode
```
3. In the settings, set `"python.analysis.diagnosticMode": "workspace"`. `"jupyter.debugJustMyCode": false`, and have several `extraPaths` to your packages' source folders, e.g.
{
    "jupyter.debugJustMyCode": false,
    "python.analysis.extraPaths": [
        "${workspaceFolder}/mybluepacks/bluepack1",
        "${workspaceFolder}/myredpacks/redpack7",
        "${workspaceFolder}/mybeloved/src",
    ],
    "python.analysis.diagnosticMode": "workspace"
}
These settings seem important to reproducing reliably, as they seem to guarantee a torrent of "files for analysis" every time the project is opened 
4. Now, every time you enter this project, pylance/pyright start to parse the entire project and analyze all files. Some of the files invariably come up with untrue diagnostics just as others have discussed.
5. A workaround is to open every file that is showing errors in the editor and wait for pylance/pyright to reanalyze that file. Sometimes it is necessary to "touch" the file e.g. add a blank somewhere in the file to trigger the reevaluation.

Thanks!
GPN

**Comment by erictraut** [comment]
@giladpn, thanks for posting. The steps you've provided above are not specific enough for me to repro.

Would you be able to create a small project (e.g. in a temporary public repo) that demonstrates the problem?

**Comment by cwells** [comment]
Hate to leave a "me too", but I started getting this today and also have no way to reproduce it, other than just noticing it pops up (usually in unrelated code) if I'm changing imports that affect typing. Making an edit to one of the affected lines of code makes it go away.

**Comment by Nicolas-Arnaise** [comment]
Hi everyone.
I hope this info can help identifying the problem:
For me it happens realy often (1 time out of 3) in subclasses with a formatter (i'm using blake and format_on_save option)
Plus, my project is realy heavy (more than 10000 lines dispatched in a lot of small files regrouped as a python module)
It looks like it occurs when the formatting takes longer to process.
 -> The problem may be caused by the file being modified while pyright is resolving classes ?

**Comment by mrckzgl** [comment]
I experience the same issue. @erictraut why is this closed?

**Comment by erictraut** [comment]
@mrckzgl, if you think you are seeing a bug, please open a new issue and provide repro steps. It's unlikely to be the same issue as the OP's issue in this thread. If you have questions about whether you are seeing a bug, open a discussion topic and provide a code sample.

**Comment by mrckzgl** [comment]
@erictraut I do think I am seeing a bug, exactly as @matt-the-midnight-hacker @danstewart @Nicolas-Arnaise @cwells @giladpn and all the others how reported here. The issue I experience has the exact same symptoms as the others describe, also it has the same workaround as posted [here](https://github.com/microsoft/pyright/issues/4518#issuecomment-1688829876). What is your definition of two issues being 'the same'? For me this qualifies enough as being the same issue. My intuition also tells me it is the same underlying problem. But out of curiosity, why do you think differently? Opening another issue for the same problem just feels like bloating up the issue tracker for no reason.

You brought up, that there are no repro steps. I agree, this is a problem, but this is unrelated from the fact that the issue exists and is confirmed by quite some users. My question was not about solving the issue, rather about why it is 'closed', which indicates it does not exist anymore and gives the problem much less visibility (for example to some how come up with repro steps).

**Comment by erictraut** [comment]
@mrckzgl, this is how I manage issues in this project. If there's nothing actionable related to an issue, it is closed.

If you are able to provide something actionable (additional details, repro steps, new variations, etc.), then I'll reopen the issue if it becomes actionable. Thanks for confirming that you're seeing a similar problem, but that does little to advance this issue.
